### DISCUSSION TITLE
What do "-with-previews -with-filters" options actually do in single-user installation?

### DISCUSSION THREAD

**Discussion by bfg01** [sumatrapdfreader/sumatrapdf#4426] [Q&A] [answered]
Do they add registry stuff to HKCU, or copy the two dlls somewhere else for SumatraPDF to read them?

Can uninstaller handle these ones as well?

**Comment by kjk** [comment, marked as answer]
It works the same in both single user and for-all-users. SumatraPDF.exe, preview.dll and filter.dll are extracted to installation directory. preview.dll and filter.dll are registered with the system using regsrv command.

The uninstaller handles are cases.

**Reply by bfg01** [reply]
I did a search for strings "regsvr" and "regsvr32" in the whole sumatrapdf Github repo, and only found some few references in issues and discussion, but nothing in the code itself.
Where could I see the parts of dll registering?

**Reply by GitHubRulesOK** [reply]
SummatraPDF does not call the registry that way as such it will talk to the registry direct from the installer so for example you would search https://github.com/search?q=repo%3Asumatrapdfreader%2Fsumatrapdf%20HKEY_CURRENT_USER&type=code

Hopefully like a well qualified "Tree Surgeon" it should (based on experience) know where to graft and prune the myriad of entries, since lopping off branches with an axe or chainsaw can lead to instability. hence all the warnings dont touch /edit  what you dont know.

**Comment by bfg01** [comment]
Slightly off-topic, but now that you mention regsvr, do you know how it actually works?

I have always been curious about this tool, which seems to not be well documented at least in the MS docs.

Does this tool write some stuff to registry (what and where)? If used with the /u option, does it completely remove all stuff from registry, even if stuff already removed? When to use regsvr32 and regsvr64?

Thanks.

**Reply by kjk** [reply]
`regsvr foo.dll` loads the dll and calls its `DllRegisterServer`. With `/u` flag it calls `DllUnregisterServer`.

What those functions do depends on the dll.

Sumatra's search filter .dll does: https://github.com/sumatrapdfreader/sumatrapdf/blob/master/src/ifilter/SearchFilterDll.cpp#L126

It calls `InstallSearchFiler` which sets a bunch of registry entries.

In our installer I don't call `regsvr`. I could but I just call `InstallSearchFiler` directly.

**Reply by GitHubRulesOK** [reply]
Hmm Is there a deliberate need for using that spelling (I know it should not matter) https://github.com/search?q=repo%3Asumatrapdfreader%2Fsumatrapdf%20InstallSearchFiler&type=code is consistent but less intuitive to a human looking for Fil **t** er

Perhaps So the two functions are separated ?

https://github.com/search?q=repo%3Asumatrapdfreader%2Fsumatrapdf%20InstallSearchFilter&type=code

**Reply by kjk** [reply]
It's a typo
